### ISSUE TITLE
Modal needs to reliably trap focus so that Carbon users don't have to

### ISSUE THREAD

**Issue by carmacleod** [carbon-design-system/carbon#1637]
Three test cases to show the problem:
1) https://codepen.io/team/carbon/pen/dZVoPj
    - open the modal
    - type tab a number of times, and focus will exit the modal (shift+tab exits modal and does not wrap)

2) https://www.carbondesignsystem.com/components/modal/code
    - open any of the modals
    - type tab until it looks as if "nothing" has focus
    - now use the up/down arrow keys... the background scrolls...

3) https://www.carbondesignsystem.com/components/modal/code
    - open any of the modals
    - type shift+tab until it looks as if "nothing" has focus (keep typing shift+tab - it does not wrap)
    - now use the up/down arrow keys... the background scrolls...

Related issues: https://github.com/IBM/carbon-components/issues/1075, https://github.com/IBM/carbon-components/issues/1474

Note that this also comes up in the forums from time to time, typically after accessibility testing.

Users of Carbon Modal eventually find out that traversing with tab or shift+tab escapes the modal.
Carbon should take care of this in a robust way, so that all Carbon users get the proper behavior for free, and they don't have to implement their own focus trap, or override Carbon code, to fix the problem.

One way to handle this is to handle the tab and shift+tab keys in a keydown event handler:
```
function confineTabToDialog(firstElement, lastElement) {    
        lastElement.keydown(function(evt) {
            if(evt.keyCode === 9 && !evt.shiftKey) {
                evt.preventDefault();
                firstElement.focus();
            }
        });
        firstElement.keydown(function(evt) {
            if(evt.keyCode === 9 && evt.shiftKey) {
                evt.preventDefault();
                lastElement.focus();
            }
        });
}
```

You can specify a way for the Carbon Modal user to tell you which elements firstElement and lastElement are. (i.e. with props or class names). This is the easiest way to get the first tabbable element and the last tabbable element in the dialog.

Alternatively, you can do the work automatically for the Carbon Modal user, and loop through the Modal's descendant elements to find the first and last tabbable element in the Modal. Note: Elements with a negative tab index are focusable, but not tabbable.

Here's some code from the [APG's Modal Dialog](http://w3c.github.io/aria-practices/#dialog_modal) [example](http://w3c.github.io/aria-practices/examples/dialog-modal/dialog.html) (found in their [utils.js](http://w3c.github.io/aria-practices/examples/js/utils.js) which is called from [dialog.js](http://w3c.github.io/aria-practices/examples/dialog-modal/js/dialog.js)):
```
aria.Utils.isFocusable = function (element) {
  if (element.tabIndex > 0 || (element.tabIndex === 0 && element.getAttribute('tabIndex') !== null)) {
    return true;
  }

  if (element.disabled) {
    return false;
  }

  switch (element.nodeName) {
    case 'A':
      return !!element.href && element.rel != 'ignore';
    case 'INPUT':
      return element.type != 'hidden' && element.type != 'file';
    case 'BUTTON':
    case 'SELECT':
    case 'TEXTAREA':
      return true;
    default:
      return false;
  }
};
```
  Note: The APG isFocusable code may be missing certain cases of OBJECT and AREA, so the jQuery code below may cover more types of element.

Here is similar code from jQuery's [:tabbable](http://api.jqueryui.com/tabbable-selector/) and [:focusable](http://api.jqueryui.com/focusable-selector/) selectors in their [jquery-ui.js](https://code.jquery.com/ui/1.12.1/jquery-ui.js) file:

```
var tabbable = $.extend( $.expr[ ":" ], {
	tabbable: function( element ) {
		var tabIndex = $.attr( element, "tabindex" ),
			hasTabindex = tabIndex != null;
		return ( !hasTabindex || tabIndex >= 0 ) && $.ui.focusable( element, hasTabindex );
	}
} );
```

```
$.ui.focusable = function( element, hasTabindex ) {
	var map, mapName, img, focusableIfVisible, fieldset,
		nodeName = element.nodeName.toLowerCase();

	if ( "area" === nodeName ) {
		map = element.parentNode;
		mapName = map.name;
		if ( !element.href || !mapName || map.nodeName.toLowerCase() !== "map" ) {
			return false;
		}
		img = $( "img[usemap='#" + mapName + "']" );
		return img.length > 0 && img.is( ":visible" );
	}

	if ( /^(input|select|textarea|button|object)$/.test( nodeName ) ) {
		focusableIfVisible = !element.disabled;

		if ( focusableIfVisible ) {

			// Form controls within a disabled fieldset are disabled.
			// However, controls within the fieldset's legend do not get disabled.
			// Since controls generally aren't placed inside legends, we skip
			// this portion of the check.
			fieldset = $( element ).closest( "fieldset" )[ 0 ];
			if ( fieldset ) {
				focusableIfVisible = !fieldset.disabled;
			}
		}
	} else if ( "a" === nodeName ) {
		focusableIfVisible = element.href || hasTabindex;
	} else {
		focusableIfVisible = hasTabindex;
	}

	return focusableIfVisible && $( element ).is( ":visible" ) && visible( $( element ) );
};
```

**Comment by asudoh** [comment]
Hi @carmacleod thank you for writing this up! Several backgrounds I have are:

1. The CodePen example needs `<input class="bx--visually-hidden">` at the bottom to make "focus loop" work
2. I believe you saw "passive modal" example. Transactional modal example sets focus to the primary button when it's launched. Application as well as passive modal can control this behavior by `data-modal-primary-focus` attribute
3. Partially connected to `2.`, "first/last focusable element" notion has been in consideration
4. The default element to put focus on launch may nees better focus indicator
5. The mechanism of "focus loop" is similar to Bootstrap, which is, trapping `focus`/`focusin` events

**Comment by joshblack** [comment]
cc @dakahn

**Comment by carmacleod** [comment]
Hi, @asudoh !

Regarding 1., what I mean is that Carbon users should not have to add a hidden input, because it is possible for the Carbon code to take care of the "focus loop" without asking every user of Carbon Modal (including CodePen example) to add a hidden input.

For 2. & 4. this issue is not about where the focus starts, it is about where it goes when the user types tab or shift+tab.

For 5. I just tried a [Bootstrap Modal](https://getbootstrap.com/docs/4.0/components/modal/)  (the one with "Launch demo modal" button) and I see that it does not work properly either. It fails in the same manner as the Carbon Modal. So the Bootstrap method of trapping focus/focusin is not the most robust method to use.

Bootstrap users are complaining about it, too:
https://github.com/twbs/bootstrap/issues/27241
https://github.com/twbs/bootstrap/issues/21661

**Comment by asudoh** [comment]
Hi @carmacleod thank you for your additional thoughts - First on bootstrap modal, when I hit "Launch demo modal" button and hit tab key, I see focus on the X icon. And then I kept hitting tab key, I see "focus loop" working. Wanted to see if it's the same as what you saw.

One thing to add to above is that after hitting tab when "Save changes" has focus, it looks as though focus is no nowhere, but you can hit tab key again to get the focus back on the X icon. That may connect to your 2. and 4. and it's where my 3. came from.

Lastly, I see a desire to supporting "focus loop" more out of the box, and we take this in consideration. One thing I'd like the team to avoid is over-engineering in this area. One quick approach may be auto-addition of focus-trap element.

**Comment by carmacleod** [comment]
To clarify, here is an example of how focus traversal in a Modal Dialog should work.
- go to the [APG Modal Dialog Example page](http://w3c.github.io/aria-practices/examples/dialog-modal/dialog.html)
- click on the button that says "Add Delivery Address"
- type the `Tab` key at least a dozen times - more if you like - until you see how the focus loop works
- now type `Shift+Tab` the same number of times
Note that focus is always within the dialog - it never leaves. The only way out (for a keyboard user) is to type `Esc` or activate the Add or Cancel button.

**Comment by carmacleod** [comment]
> One quick approach may be auto-addition of focus-trap element.

Please don't do this. It's a hack, and it doesn't completely fix the problem.  ;)

> it looks as though focus is nowhere, but you can hit tab key again to get the focus back on the X icon

This is incorrect behavior.

**Comment by asudoh** [comment]
Thank you for sharing an example @carmacleod BTW! While I see a desire for sophisticated "focus loop" support in our modal, https://material.io/develop/web/components/dialogs/#handling-focus-trapping well explains that library-side implementation of this is quite complex. Given this, as we consider the approach seen in http://w3c.github.io/aria-practices/examples/dialog-modal/dialog.html, providing it as an example or an add-on would be one option. Creating an API endpoint so that application can use a dedicated library for "focus loop", e.g. [`focus-trap`](https://www.npmjs.com/package/focus-trap) library, may be another option.

**Comment by carmacleod** [comment]
Very interesting, @asudoh - thanks.

So, I think you could go one better than Material Components if you create the focus trap by default, and provide an API called `disableFocusTrap` that has a comment saying something like, "Use this to disable the focus trap if you are using another way to trap focus, like [`inert`](https://github.com/Fyrd/caniuse/issues/3090) _(which would/will solve this problem completely, by the way ;)_ or some other framework."

The current Material Components method allows accessibility failure by default, and hopes that the developer notices the focus trap api and understands why they need to use it (or find an alternative).
Even [their own dialog example](https://material-components.github.io/material-components-web-catalog/#/component/dialog) does not implement the focus trap.

In case it's helpful, here is the [focus-trap code](https://github.com/davidtheclark/focus-trap) that they are using, and I see that this code uses [yet another implementation of tabbable](https://github.com/davidtheclark/tabbable).

If you decide to implement it the same way as Material Components (i.e. trapFocus and releaseFocus), I would be ok with that, as long as your examples use it, and you have a nice obvious comment that says a focus trap is required for accessibility.  :)

By the way, I just noticed that your React Modal does not have `aria-modal="true"` - that's a [bug](https://github.com/IBM/carbon-components-react/issues/1752).

**Comment by joshblack** [comment]
> The current Material Components method allows accessibility failure by default, and hopes that the developer notices the focus trap api and understands why they need to use it (or find an alternative). Even their own dialog example does not implement the focus trap.

Even when looking for it on the page, I had a hard time locating the focus trap guidance, unfortunately. 

I do think there is a good opportunity, at least in React, to provide this stuff by default. For Vanilla.js, would love to find a way to support it by default with an opt-out. As @carmacleod mentioned, we should try and aim for accessible by default. What kinds of things can we offer that could work here to accomplish that?

**Comment by dakahn** [comment]
addressed in [this pull request](https://github.com/IBM/carbon-components-react/pull/2096) 👍
